File: tests/whisper_ffi/WhisperProbe.kt
package whisperprobe

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import uniffi.cera_ffi.FfiException
import uniffi.cera_ffi.FfiWhisperModel
import uniffi.cera_ffi.whisperDefaultTranscribeOpts
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.file.Files
import java.nio.file.Path

// This helper is also the application-facing example in the Whisper guide.
suspend fun transcribeRecording(
    modelPath: String,
    pcm16kMono: List<Float>,
): String {
    return withContext(Dispatchers.IO) {
        FfiWhisperModel.fromFile(modelPath).use { model ->
            val opts = whisperDefaultTranscribeOpts().copy(language = "en")
            model.transcribeAsync(pcm16kMono, opts)
        }
    }
}

fun main(args: Array<String>) =
    runBlocking {
        val directory = Path.of(args.single())
        val raw = Files.readAllBytes(directory.resolve("audio.f32"))
        check(raw.isNotEmpty() && raw.size % 4 == 0)
        val buffer = ByteBuffer.wrap(raw).order(ByteOrder.LITTLE_ENDIAN)
        val pcm = List(raw.size / 4) { buffer.float }
        val bytes = Files.readAllBytes(directory.resolve("a.gguf"))
        FfiWhisperModel.fromBytes(bytes).use { a ->
            bytes.fill(0)
            val copy = directory.resolve("kotlin-owned.gguf")
            Files.copy(directory.resolve("b.gguf"), copy)
            FfiWhisperModel.fromFile(copy.toString()).use { b ->
                Files.delete(copy)
                check(a.isMultilingual() && !b.isMultilingual())
                check(a.languages().size == 100 && a.languages().take(2) == listOf("en", "zh"))
                val defaults = whisperDefaultTranscribeOpts()
                check(defaults.language == null && !defaults.translate && !defaults.timestamps)
                check(defaults.maxTokens == 448u && defaults.temperature == 0f)
                val opts = defaults.copy(language = "en", maxTokens = 3u)
                check(a.transcribe(pcm, opts) == "aaa")
                check(a.transcribe(emptyList(), null).isEmpty())
                val short = opts.copy(maxTokens = 2u)
                val first = async { a.transcribeAsync(pcm, opts) }
                val second = async { a.transcribeAsync(pcm, short) }
                val third = async { b.transcribeAsync(pcm, short) }
                check(listOf(first.await(), second.await(), third.await()) == listOf("aaa", "aa", "bb"))
                check(a.transcribe(pcm, opts.copy(maxTokens = null, temperature = null)) == "a".repeat(13))
                check(transcribeRecording(directory.resolve("a.gguf").toString(), pcm) == "a".repeat(13))
            }
        }
        try {
            FfiWhisperModel.fromBytes(byteArrayOf(1, 2)).use { error("Malformed GGUF accepted") }
        } catch (_: FfiException.Backend) {
        }
        val missing = directory.resolve("missing.gguf")
        check(!Files.exists(missing))
        try {
            FfiWhisperModel.fromFile(missing.toString()).use { error("Missing GGUF accepted") }
        } catch (_: FfiException.Backend) {
        }
        println(
            """{"cases":["bytes-owned","file-owned","languages","defaults","sync","async-shared","async-distinct","empty","malformed","missing","recording-example"],"text":["aaa","aa","bb"]}""",
        )
    }

File: tests/whisper_ffi/WhisperProbe.swift
import Cera
import Foundation

// This helper is also the application-facing example in the Whisper guide.
func transcribeRecording(modelPath: String, pcm16kMono: [Float]) async throws -> String {
    let model = try await Task.detached {
        try FfiWhisperModel.fromFile(path: modelPath)
    }.value
    var opts = whisperDefaultTranscribeOpts()
    opts.language = "en"
    return try await model.transcribeAsync(pcm: pcm16kMono, opts: opts)
}

@main
struct WhisperProbe {
    static func main() async throws {
        let directory = URL(fileURLWithPath: CommandLine.arguments[1])
        let raw = try Data(contentsOf: directory.appendingPathComponent("audio.f32"))
        precondition(raw.count > 0 && raw.count % 4 == 0)
        let pcm: [Float] = stride(from: 0, to: raw.count, by: 4).map { offset in
            let bits = (0..<4).reduce(UInt32(0)) { $0 | UInt32(raw[offset + $1]) << (8 * $1) }
            return Float(bitPattern: bits)
        }
        var bytes = try Data(contentsOf: directory.appendingPathComponent("a.gguf"))
        let a = try FfiWhisperModel.fromBytes(bytes: bytes)
        bytes.resetBytes(in: 0..<bytes.count)
        let copy = directory.appendingPathComponent("swift-owned.gguf")
        try FileManager.default.copyItem(at: directory.appendingPathComponent("b.gguf"), to: copy)
        let b = try FfiWhisperModel.fromFile(path: copy.path)
        try FileManager.default.removeItem(at: copy)
        precondition(a.isMultilingual() && !b.isMultilingual())
        precondition(a.languages().count == 100 && Array(a.languages().prefix(2)) == ["en", "zh"])
        var opts = whisperDefaultTranscribeOpts()
        precondition(opts.language == nil && !opts.translate && !opts.timestamps)
        precondition(opts.maxTokens == 448 && opts.temperature == 0)
        opts.language = "en"
        opts.maxTokens = 3
        let sync = try a.transcribe(pcm: pcm, opts: opts)
        precondition(sync == "aaa")
        let empty = try a.transcribe(pcm: [], opts: nil)
        precondition(empty.isEmpty)
        var short = opts
        short.maxTokens = 2
        let fullOptions = opts
        let shortOptions = short
        async let first = a.transcribeAsync(pcm: pcm, opts: fullOptions)
        async let second = a.transcribeAsync(pcm: pcm, opts: shortOptions)
        async let third = b.transcribeAsync(pcm: pcm, opts: shortOptions)
        let results = try await [first, second, third]
        precondition(results == ["aaa", "aa", "bb"])
        // None/null option fields use core defaults; context caps this tiny model.
        opts.maxTokens = nil
        opts.temperature = nil
        let defaults = try a.transcribe(pcm: pcm, opts: opts)
        precondition(defaults == String(repeating: "a", count: 13))
        let recording = try await transcribeRecording(
            modelPath: directory.appendingPathComponent("a.gguf").path, pcm16kMono: pcm)
        precondition(recording == defaults)
        do {
            _ = try FfiWhisperModel.fromBytes(bytes: Data([1, 2]))
            preconditionFailure("Malformed GGUF accepted")
        } catch FfiError.Backend { }
        let missing = directory.appendingPathComponent("missing.gguf").path
        precondition(!FileManager.default.fileExists(atPath: missing))
        do {
            _ = try FfiWhisperModel.fromFile(path: missing)
            preconditionFailure("Missing GGUF accepted")
        } catch FfiError.Backend { }
        print(#"{"cases":["bytes-owned","file-owned","languages","defaults","sync","async-shared","async-distinct","empty","malformed","missing","recording-example"],"text":["aaa","aa","bb"]}"#)
    }
}
